src/rpc/rawtransaction.cpp

```diff
@@ -916,13 +917,17 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
 
     TxValidationState state;
     bool test_accept_res;
+    CAmount fee; // To return transaction fee from AcceptToMemoryPool
     {
         LOCK(cs_main);
         test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true);
+            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true, &fee);
     }
     result_0.pushKV("allowed", test_accept_res);
-    if (!test_accept_res) {
+    if (test_accept_res) {
```

Comment:
Could you elaborate on the consideration behind only passing back fee rate if the transaction is accepted? Late to the party, so I might have missed it if this was asked earlier by a different reviewer.

I think we should still pass back fee even if rejected, because In validation.h, we can see that one potential reason of rejection is`TX_MEMPOOL_POLICY`, which includes reasons such as mempool min fee not met, min relay fee not met, etc (in `CheckFeeRate`). 

This RPC is also potentailly improved to process more than one transaction in the future, which means that more than one trasactions could fail. At that point, it would be helpful for users to see directly the transactions that are not accepted & the exact fees they chose for the corresponding transactions.

Reply:
The rationale behind not returning fees in case of rejection was mempool policy is not the only reason why a transaction might be rejected. In case of ill-formed transactions, returning a fee might cause a problem because it might try to return something that doesn't make sense. That might add up confusion about the result than help.

On the other hand, `CheckFeeRate` already returns both the `package fee` and `min_fee*` as an error message. So that information is already available. 

So it felt redundant to try to return fee in case of failure in the RPC call and handle all the edge cases. If it needs to be done I think it's better to do it in a separate PR and it will need some added testings too to ensure what is returned actually make sense.

Reply:
>  If I am not mistaken then it will return 0 fee for ill-formed transactions.
@rajarshimaitra would you mind pointing me to the code to look more into it? I did some searching myself but wasn't able to find this.

Reply:
Check here.

https://github.com/bitcoin/bitcoin/blob/205b87d2f6bd01285de50ba742e32e4ab1298b13/src/validation.cpp#L688-L691

if `CheckTxInputs` fails it will return `nFess` will be 0. Look inside `CheckTxInputs` to see what makes it fail.